src/core/stdcpp/exception.d

```diff
@@ -6,95 +6,111 @@
  * Authors:   $(HTTP digitalmars.com, Walter Bright)
+ *            Manu Evans
  * Source:    $(DRUNTIMESRC core/stdcpp/_exception.d)
  */
 
 module core.stdcpp.exception;
 
-extern (C++, "std"):
+import core.stdcpp.xutility : __cplusplus, CppStdRevision;
 
 version (CppRuntime_DigitalMars)
-{
-    import core.stdcpp.typeinfo;
+    version = GenericBaseException;
+version (CppRuntime_Gcc)
+    version = GenericBaseException;
+version (CppRuntime_Clang)
+    version = GenericBaseException;
 
-    alias void function() unexpected_handler;
-    unexpected_handler set_unexpected(unexpected_handler f) nothrow;
-    void unexpected();
-
-    alias void function() terminate_handler;
-    terminate_handler set_terminate(terminate_handler f) nothrow;
-    void terminate();
+extern (C++, "std"):
+@nogc:
 
-    bool uncaught_exception();
+///
+alias terminate_handler = void function() nothrow;
+///
+terminate_handler set_terminate(terminate_handler f) nothrow;
+///
+terminate_handler get_terminate() nothrow;
+///
+void terminate() nothrow;
 
-    class exception
-    {
-        this() nothrow { }
-        this(const exception) nothrow { }
-        //exception operator=(const exception) nothrow { return this; }
-        //virtual ~this() nothrow;
-        void dtor() { }
-        const(char)* what() const nothrow;
-    }
+static if (__cplusplus < CppStdRevision.cpp17)
+{
+    ///
+    alias unexpected_handler = void function();
+    ///
+    deprecated unexpected_handler set_unexpected(unexpected_handler f) nothrow;
+    ///
+    deprecated unexpected_handler get_unexpected() nothrow;
+    ///
+    deprecated void unexpected();
+}
 
-    class bad_exception : exception
-    {
-        this() nothrow { }
-        this(const bad_exception) nothrow { }
-        //bad_exception operator=(const bad_exception) nothrow { return this; }
-        //virtual ~this() nothrow;
-        override const(char)* what() const nothrow;
-    }
+static if (__cplusplus < CppStdRevision.cpp17)
+{
+    ///
+    bool uncaught_exception() nothrow;
 }
-else version (CppRuntime_Gcc)
+else static if (__cplusplus == CppStdRevision.cpp17)
 {
-    alias void function() unexpected_handler;
-    unexpected_handler set_unexpected(unexpected_handler f) nothrow;
-    void unexpected();
-
-    alias void function() terminate_handler;
-    terminate_handler set_terminate(terminate_handler f) nothrow;
-    void terminate();
-
-    pure bool uncaught_exception();
+    ///
+    deprecated bool uncaught_exception() nothrow;
+}
+static if (__cplusplus >= CppStdRevision.cpp17)
+{
+    ///
+    int uncaught_exceptions() nothrow;
+}
 
+version (GenericBaseException)
+{
+    ///
     class exception
     {
-        this();
-        //virtual ~this();
-        void dtor1();
-        void dtor2();
-        const(char)* what() const;
-    }
+    @nogc:
+        ///
+        this() nothrow {}
+        ///
+        ~this() nothrow {} // HACK: this should extern, but then we have link errors!
```

Comment:
How is that supposed to work?

Reply:
it reserves the appropriate position in the vtable.

Reply:
D can't throw C++ exceptions, and also doesn't seem to be able to catch C++ exceptions either.